rascal-lsp/src/main/rascal/lang/rascal/lsp/IDECheckerWrapper.rsc

```diff
@@ -0,0 +1,134 @@
+import lang::rascalcore::check::Checker;
+import lang::rascalcore::check::ModuleLocations;
+
+@synopsis{
+    This function is a wrapper for the Rascal type checker, taking into account that dependencies
+    may be open in the current workspace. It takes an overapproximation of the import/extend graph,
+    and calls the type checker on (transitive) dependencies first.
+
+    This function must only be used in an IDE context, as this is the only situation in which non-lib
+    source locations can occur in the `libs` entry of the PathConfig of a project. Note that for `lib`
+    locations, the type checker uses `tpl` files that are packaged with libraries.
+}
+list[ModuleMessages] checkFile(loc l, start[Module](loc file) getParseTree, PathConfig(loc file) getPathConfig) 
+    = job("Rascal check", list[ModuleMessages](void(str, int) step) {
+    checkForImports = [getParseTree(l)];
+    checkedForImports = {};
+    initialProject = inferProjectRoot(l);
+
+    rel[loc, loc] dependencies = {};
+    
+
+    step("Building dependency graph", 50);
+    while (tree <- checkForImports) {
+        currentSrc = tree.src.top;
+        currentProject = inferProjectRoot(currentSrc);
+        for (i <- tree.top.header.imports) {
+            try {
+                ml = locateRascalModule("<i.\module>", getPathConfig(currentProject), getPathConfig);
+                if (ml.extension == "rsc", mlpt := getParseTree(ml), mlpt.src.top notin checkedForImports) {
+                    checkForImports += mlpt;
+                    dependencies += <currentProject, inferProjectRoot(mlpt.src.top)>;
+                }
+            } catch _: {
+                ;// Continue
+            }
+        }
+        checkedForImports += currentSrc;
+        checkForImports -= tree;
+    }
+    if (<p,p> <- (dependencies - ident(carrier(dependencies)))+) {
+        return [program(l, error("Detected cyclic dependencies between projects. Fix your project setup.", l))];
+    }
+    modulesPerProject = classify(checkedForImports, loc(loc l) {return inferProjectRoot(l);});
+    msgs = [];
+    for (project <- reverse(order(dependencies)), project in modulesPerProject, project != initialProject) {
+        step("Checking project `<project.file>`", 1);
+        msgs += check([*modulesPerProject[project]], rascalCompilerConfig(getPathConfig(project))); 
+    }
+    step("Checking <l>", 1);
+    msgs += check([l], rascalCompilerConfig(getPathConfig(initialProject)));
+    return dup(msgs);
+});
+
+loc locateRascalModule(str fqn, PathConfig pcfg, PathConfig(loc file) getPathConfig) {
+    fileName = makeFileName(fqn);
+    // Check the source directories
+    for (dir <- pcfg.srcs, fileLoc := dir + fileName, exists(fileLoc)) {
+        return fileLoc;
+    }
+    // Check the source directories of libraries that are currently open in VS Code
+    if (lib <- pcfg.libs, lib.scheme != "lib", dir <- getPathConfig(inferProjectRoot(lib)).srcs, fileLoc := dir + fileName, exists(fileLoc)) {
```

Comment:
in the rename function we explicitly have a list of current project in the workspace that is passed in. should we also do that? so that we don't try to read from non workspace locations? (especially as the `lib` scheme is soon dissapearing and getting replaced by for example `jar+file` and `mvn`)

Reply:
That sounds like a good idea